api/v1beta1/machinedeployment_webhook.go

```diff
@@ -213,3 +261,105 @@ func (m *MachineDeployment) validate(old *MachineDeployment) error {
+//     disruptive for existing Machines and if 1 is outside the (min size, max size) range the autoscaler won't take
+//     control.
+//
+// Notes:
+//   - While the min size and max size annotations of the autoscaler provide the best UX, other autoscalers can use the
+//     DefaultReplicasAnnotation if they have similar use cases.
+func calculateMachineDeploymentReplicas(oldMD *MachineDeployment, newMD *MachineDeployment) (int32, error) {
+	// If replicas is already set => Keep the current value.
+	if newMD.Spec.Replicas != nil {
+		return *newMD.Spec.Replicas, nil
+	}
+
+	// Default replica annotation is set => Use the annotation value.
+	if defaultReplicasString, ok := newMD.Annotations[DefaultReplicasAnnotation]; ok {
+		defaultReplicas, err := strconv.ParseInt(defaultReplicasString, 10, 32)
+		if err != nil {
+			return 0, errors.Wrapf(err, "failed to caculate MachineDeployment replicas value: could not parse the value of the %q annotation", DefaultReplicasAnnotation)
+		}
+		return int32(defaultReplicas), nil
+	}
+
+	// If both autoscaler annotations are set, use them to calculate the default value.
+	minSizeString, hasMinSizeAnnotation := newMD.Annotations[autoscalerMinSize]
+	maxSizeString, hasMaxSizeAnnotation := newMD.Annotations[autoscalerMaxSize]
+	if hasMinSizeAnnotation && hasMaxSizeAnnotation {
+		minSize, err := strconv.ParseInt(minSizeString, 10, 32)
+		if err != nil {
+			return 0, errors.Wrapf(err, "failed to caculate MachineDeployment replicas value: could not parse the value of the %q annotation", autoscalerMinSize)
+		}
+		maxSize, err := strconv.ParseInt(maxSizeString, 10, 32)
+		if err != nil {
+			return 0, errors.Wrapf(err, "failed to caculate MachineDeployment replicas value: could not parse the value of the %q annotation", autoscalerMaxSize)
+		}
+
+		// If it's a new MachineDeployment => Use the min size.
+		// Note: This will result in a scale up to get into the range where autoscaler takes over.
+		if oldMD == nil {
+			return int32(minSize), nil
+		}
+
+		// Otherwise we are handing over the control for the replicas field for an existing MachineDeployment
+		// to the autoscaler.
+
+		switch {
+		// If the old MachineDeployment doesn't have replicas set => Use the min size.
+		// Note: As defaulting always sets the replica field, this case should not be possible
+		// We only have this handling to be 100% safe against panics.
+		case oldMD.Spec.Replicas == nil:
+			return int32(minSize), nil
+		// If the old MachineDeployment replicas are lower than min size => Use the min size.
+		// Note: This will result in a scale up to get into the range where autoscaler takes over.
+		case *oldMD.Spec.Replicas < int32(minSize):
+			return int32(minSize), nil
+		// If the old MachineDeployment replicas are higher than max size => Use the max size.
+		// Note: This will result in a scale down to get into the range where autoscaler takes over.
+		case *oldMD.Spec.Replicas > int32(maxSize):
+			return int32(maxSize), nil
```

Comment:
it seems like we should be emitting a log message or something here so that a user would know the value is changed from the previous?

Reply:
We can. But I'm not sure if a user will notice a log line in the controller.

If we want we could return a webhook warning with a future CR version (probably v0.15 and v0.16)

Reply:
i'm not sure a user would notice the log line either, but i think having some sort of audit trail is nice. presumably this is the type of thing that would be found in a debrief about "why did MachineDeployment X suddenly change replica count".

> Just to clarify. The value is not directly changed, it is first unset by the client and then the default will be the max size when running defaulting after the unset (all happens in one call)

ack, i was just thinking about the cases where an old MD is being included in autoscaling.

this isn't a blocker for me if you don't think it's a high value warning.

Reply:
> i'm not sure a user would notice the log line either, but i think having some sort of audit trail is nice.

Sounds good, will add the log lines.